### DISCUSSION TITLE
Accordion

### DISCUSSION THREAD

**Discussion by bameyrick** [ONSdigital/design-system#1377] [Feedback]
Use this place to discuss the [Accordion component](https://ons-design-system.netlify.com/components/accordion) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by dudelmeister** [comment]
Accordion needs bottom padding on the content div

**Comment by jrbarnes9** [comment]
> Accordion needs bottom padding on the content div

PR -> https://github.com/ONSdigital/pattern-library-v2/pull/112

**Comment by bameyrick** [comment]
When a [details](#18) component is open, it's button text does not change to 'Close'

**Comment by jrbarnes9** [comment]
Accordions alignment broken on small breakpoint
![Screenshot 2020-11-03 at 13 26 39](https://user-images.githubusercontent.com/43346934/97990681-3b4d5f80-1dd8-11eb-8140-421545f80b22.png)

**Comment by trevorsaint** [comment]
Accordions break visually if using it without show/hide buttons. I have removed one as an example.

<img width="783" alt="Screenshot 2021-01-18 at 11 52 10" src="https://user-images.githubusercontent.com/4989027/104912215-e73c2b00-5983-11eb-9f3b-e80caa151e78.png">

**Reply by jrbarnes9** [reply]
It may be a case that the buttons should be mandatory. Component needs to be retested with DAC because the original evidence for the buttons cannot be found. It is believed the buttons to open/close the accordions were added as a visual cue required for users of speech recognition software.

**Reply by jrbarnes9** [reply]
To be DAC tested in May 2022 https://github.com/onsdigital/design-team/issues/10

**Comment by trevorsaint** [comment]
A visual improvement would be to indent the content, so visually it aligns to the heading.

An indent of 27px or 1.5rem on the content would achieve this.

### Currently
<img width="633" alt="current" src="https://user-images.githubusercontent.com/4989027/117452131-50e44300-af3b-11eb-9b0f-fc66c0c79422.png">

### Suggested
<img width="663" alt="suggested" src="https://user-images.githubusercontent.com/4989027/117452167-5b064180-af3b-11eb-9281-ceb20a3477f5.png">

**Comment by JoelClemence** [comment]
I have attempted to use this in my Flask app but this component unfortunately does not work in Jinja2 (3.0.3):

```jinja2.exceptions.TemplateRuntimeError: No filter named 'setAttributes' found.```

**Reply by jrbarnes9** [reply]
Raised #1786

**Reply by boxadesign** [reply]
@JoelClemence We will need to provide documentation for this or perhaps refactor to not require this filter. But you will need to add the filter to your app as done here:

https://github.com/ONSdigital/eq-questionnaire-runner/blob/567f85c5bcaf926a0f9534fa8226b06305d2ea70/app/jinja_filters.py#L181

**Comment by jrbarnes9** [comment]
Accordion collapsibles of the simple variant are not in the tab order

**Reply by jrbarnes9** [reply]
Raised #2147

**Comment by Al-Davies** [comment]
UR findings on accordions - 
tested as part of eQ business survey usability testing Oct 21

- users expected to be able to add their information into the survey introduction page - they expected that by clicking on the show all / show buttons this would reveal a list of questions with a box to input an answer
- users realised on trying to do this that only information was held here, quickly realising they had to use the start survey button
- users using a mixture of show all and show buttons - actions of doing so matched their expectations
- where accordians hold information, users may be less likely to look at these as they become more familiar with the product - this may need consideration if important information is added here that needs to be communicated to users

**Comment by sanjeevz3009** [comment]
### Accordion Aria Labels For Button Text

```
"allButton": {
            "open": "Show all",
            "close": "Hide all"
        },
```

Currently, if we have more than one accordion component on the page with the same button text for opening and closing the accordion component, there will be accessibility issues:

If we have more than one accordion component, both with a `show all/ hide all` button text (leaving the button texts as it is), especially when we want to be consistent around the accordion `allButton` property config, where we want the same open and close button text across the page for multiple accordion components, we run into some accessibility issues such as:
- Screen reader ambiguity: Screen reader users often navigate by button list or tabbing between interactive elements.
  - Users might activate the wrong one or waste time guessing.
- Blind or low-vision users lose context.
- Show all/ hide all buttons will be in place when there are multiple accordion components displayed on the same page. They sound identical, so the user has no idea which controls what, and they will have no context.
  - E.g. We can have one accordion component containing glossary terms and another one containing something else unrelated to glossary terms. There can be more than one or two accordions on the same page. All are using the same button text, and this is when accessibility issues arise, such as Screen reader ambiguity.
  - Having the ability to set aria labels for each accordion component will benefit massively in these scenarios and address the accessibility issues mentioned here.

**Comment by precious-onyenaucheya-ons** [comment]
Thanks for raising this. The Accordion component already includes an attributes parameter, which can be used to pass in additional HTML attributes for the button.

For example:

```
"allButton": {
  "open": "Show all",
  "close": "Hide all",
  "attributes": {
    "aria-label": "Show or hide all glossary terms"
  }
}
```


This way, you can give each accordion’s button a unique aria-label (or other identifying attributes) to help screen readers distinguish between multiple accordions on the same page.

**Comment by sanjeevz3009** [comment]
@precious-onyenaucheya-ons 

Thank you for the response. This is really helpful, Precious. Maybe some additional guidance on this or an example in the [Accordion Macro Options Table](https://service-manual.ons.gov.uk/design-system/components/accordion#example.example-accordion.nunjucks-macro-options) will be helpful, as accessibility is an important aspect that could be explicit here.

I do have some concerns with this approach, as it partly solves the problem, and there are still some gaps:
```
"allButton": {
  "open": "Show all",
  "close": "Hide all",
  "attributes": {
    "aria-label": "Show or hide all glossary terms"
  }
}
```

This does not clarify to the user whether their action will show or hide. They would need to work out using the combination of the `button` label, which may say `Show all`, and then they may be able to determine which is not ideal.

**Comment by precious-onyenaucheya-ons** [comment]
Thanks, @sanjeevz3009. That’s a good point about the current approach still not making it clear whether the button will show or hide the content. We’ll raise a ticket to look into this and find a clearer solution. I’ll let you know once it is done.

**Reply by sanjeevz3009** [reply]
@precious-onyenaucheya-ons Thank you!

**Reply by MagdalenaLarge** [reply]
Hello @sanjeevz3009 ,
Thank you for your patience. Our team was able to address the issue you raised and it was implemented during our latest release [73.0.4](https://github.com/ONSdigital/design-system/releases/tag/73.0.4).
If you have any additional questions, feel free to reach out to us.

**Reply by sanjeevz3009** [reply]
@MagdalenaLarge Thanks!
